File: client/src/fight/trace.ts
// The fight trace as `crates/lab/src/trace.rs` writes it.
//
// This file is the reader half of a two-file contract and it is deliberately
// not the worker protocol: no wasm, no ABI, no generated header. The trace
// exists so somebody can look at an articulated fight before anything else about
// it is calibrated, and it can be deleted in the commit that lands the real
// pose channel.
//
// Every number that came out of the simulation is a raw 16.16 fixed-point
// integer. Nothing in here divides one; the views do, once, at the point they
// turn a world coordinate into a pixel. That keeps exact comparisons -- "is this
// energy above the floor", "did this tip move at all" -- on the integers the
// simulation actually decided them on.

export const TRACE_SCHEMA = "arpg-fight-trace-4";

export type V3 = readonly [number, number, number];
/** `[index, generation]`, as `EntityId` is spelled on the Rust side. */
export type EntityKey = readonly [number, number];

export interface Segment {
  readonly hilt: V3;
  readonly tip: V3;
  readonly radius: number;
}

export interface ShieldFace {
  readonly centre: V3;
  readonly normal: V3;
  readonly halfWidth: number;
  readonly halfHeight: number;
  readonly thickness: number;
}

export interface Region {
  readonly lower: V3;
  readonly upper: V3;
  readonly radius: number;
  readonly present: boolean;
}

/** One live articulated arrow. Slot plus generation is its stable identity. */
export interface Projectile {
  readonly id: EntityKey;
  readonly owner: EntityKey;
  readonly position: V3;
  readonly velocity: V3;
  readonly radius: number;
  readonly remainingRange: number;
}

export interface Arm {
  readonly hand: V3;
  /** Relative to the body origin, per `PosedArm::velocity`. */
  readonly vel: V3;
  /** Where the actuator is trying to put the hand, world space. */
  readonly target: V3;
  readonly fatigue: number;
}

export interface Pose {
  readonly id: EntityKey;
  readonly body: V3;
  /** Binary angle, 65536 to the turn. */
  readonly yaw: number;
  readonly vel: V3;
  readonly arms: readonly [Arm, Arm];
  readonly weapons: readonly [Segment | null, Segment | null];
  readonly shield: ShieldFace | null;
  readonly regions: readonly Region[];
  readonly integrity: readonly number[];
  readonly wound: readonly number[];
  readonly blood: number;
  readonly shock: number;
  readonly severed: number;
  readonly equipmentMask: number;
  readonly intent: "hold" | "attack" | "flee";
  readonly target: EntityKey | null;
  readonly hints: readonly [number, number];
}

export interface Contact {
  readonly a: EntityKey;
  readonly aSlot: number;
  readonly b: EntityKey;
  readonly bSlot: number;
  /** Index into `Trace.contactKinds`. */
  readonly kind: number;
  /** Index into `Trace.regionNames`, or `Trace.noRegion`. */
  readonly region: number;
  readonly point: V3;
  readonly normal: V3;
  /**
   * The five columns a live fight does not have, and null is what says so.
   *
   * `ContactResolution` carries a velocity and an impulse per side and a group
   * alpha; the published 32-word event row carries none of them -- the alpha
   * because it is a solver search result, the other four because nothing had
   * asked for them when the row was frozen. A trace, which is written from the
   * whole struct, fills all five. Growing the row would move
   * `COMBAT_EVENT_LAYOUT_VERSION` and `ARTICULATED_STREAM_DIGEST`, so it belongs
   * to a session that wants them rather than to the one that noticed.
   *
   * Nullable rather than zero-filled, because a zero closing speed is a
   * measurement -- two colliders that met while moving together -- and a reader
   * cannot tell it from an absence.
   */
  readonly velocityA: V3 | null;
  readonly velocityB: V3 | null;
  readonly impulseA: V3 | null;
  readonly impulseB: V3 | null;
  readonly toi: number;
  readonly group: number;
  readonly alpha: number | null;
  /** Per **group**, not per contact: `closure_energy` over every collider in
   *  the time-of-impact group, bodies' own translational energy included, and
   *  copied unchanged into every row of that group. Never compare it to
   *  `contactEnergyFloor` -- use {@link share}. */
  readonly groupBefore: number;
  readonly groupAfter: number;
  readonly groupDissipated: number;
  readonly cut: number;
  readonly thrust: number;
  /**
   * Everything the edge and the point did not take.
   *
   * **Not purely inert, and the name is older than the mechanic.** Since
   * combat-arms-05 the sim splits this remainder again, into a *crushing*
   * channel that wounds and a residual that does not, and the event layout has
   * three channel words rather than four -- so what arrives here is
   * `crush + pressure`. A wooden club's whole blow is in this number and it
   * does real damage; a blade's is the energy floor and does none. The two
   * cannot be told apart from the published words alone, which is a known gap
   * and a layout change to close.
   */
  readonly pressure: number;
  readonly deflected: number;
  readonly severed: boolean;
}

/**
 * The energy this one contact was allocated out of its group's dissipation.
 *
 * **The only quantity `CONTACT_ENERGY_FLOOR` is charged against.** `channels()`
 * deducts the floor from this share and splits what is left between cut,
 * thrust and crush, and the publisher folds crush back into `pressure` -- so
 * the three published channels still sum back to the share exactly, which is
 * what makes this function correct and what keeps a crushing blow's ring the
 * size the blow actually was. Zero for weapon-weapon and weapon-shield rows,
 * which have no wound channel and pay no floor.
 */
export function share(contact: Contact): number {
  return contact.cut + contact.thrust + contact.pressure;
}

/**
 * How fast the two colliders were closing at the fact, raw units per tick, or
 * null when the source did not publish the two velocities.
 */
export function closureSpeed(contact: Contact): number | null {
  const { velocityA, velocityB } = contact;
  if (velocityA === null || velocityB === null) return null;
  return length(sub(velocityA, velocityB));
}

export interface Frame {
  readonly t: number;
  readonly poses: readonly Pose[];
  readonly projectiles: readonly Projectile[];
  readonly contacts: readonly Contact[];
  /** Heroes, then Monsters. */
  readonly health: readonly [number, number];
}

export interface Carried {
  readonly action: string;
  readonly binding: string;
  readonly mass: number;
  readonly balance: number;
  readonly geometry: "segment" | "shield";
  readonly length?: number;
  readonly radius?: number;
  readonly halfWidth?: number;
  readonly halfHeight?: number;
  readonly thickness?: number;
}

export interface BodyInfo {
  readonly index: number;
  readonly kind: string;
  readonly faction: string;
  readonly anatomy: {
    readonly standingHeight: number;
    readonly shoulderHeight: number;
    readonly shoulderHalfWidth: number;
    readonly armLength: number;
    readonly handRadius: number;
  };
  readonly carried: readonly (Carried | null)[];
}

export interface Trace {
  readonly schema: string;
  /** Raw units in one world unit. 65536, carried rather than assumed. */
  readonly one: number;
  readonly scenario: string;
  readonly mirrored: boolean;
  /** Null for a mirrored run, which is a run of a different scenario. */
  readonly fingerprint: string | null;
  readonly seed: number;
  /**
   * What drove each side.
   *
   * Schema 2 carried one `script`, because until v2-19 every trace put the same
   * script on both bodies. A learned fight has a checkpoint on the Fighter and a
   * script on the Brute, and a header that named only one of them would leave a
   * reader to guess which -- so both are published and the schema moved rather
   * than the field quietly changing meaning.
   */
  readonly heroes: string;
  readonly monsters: string;
  /** SHA-256 of the checkpoint that drove the learned side, or null. */
  readonly checkpoint: string | null;
  readonly outcome: string;
  readonly timedOut: boolean;
  readonly ticks: number;
  readonly maxTicks: number;
  readonly arena: readonly [number, number];
  readonly frameCount: number;
  readonly truncated: boolean;
  readonly impactThreshold: number;
  readonly contactEnergyFloor: number;
  readonly regionNames: readonly string[];
  readonly hintNames: readonly string[];
  readonly contactKinds: readonly string[];
  readonly bodySlot: number;
  readonly noRegion: number;
  readonly bodies: readonly BodyInfo[];
  readonly frames: readonly Frame[];
}

/**
 * Indexed access that fails loudly.
 *
 * `noUncheckedIndexedAccess` is on, and the honest answer to "what if frame 4000
 * is missing" is that the file is malformed and every number after it is a lie.
 * A `?? fallback` would draw that lie.
 */
export function at<T>(list: readonly T[], index: number): T {
  const value = list[index];
  if (value === undefined) throw new Error(`trace index ${index} is out of range`);
  return value;
}

/**
 * There is no recording at this URL, which is not the same as a broken one.
 *
 * Recorded fights are a development fixture -- `.gitignore` excludes
 * `web/fight*.json`, `npm run trace` writes them, and the production bundle
 * carries none of them -- so "absent" is the *expected* answer in a shipped
 * build and in a fresh clone alike. A viewer that reported it the way it
 * reports a corrupt file would tell a reader the application is broken when it
 * is not, which is why absence is a type here rather than one more message a
 * caller has to pattern-match a string out of.
 */
export class TraceUnavailable extends Error {
  /** The status that carried the refusal; 200 when a host answered with a page. */
  readonly status: number;

  constructor(message: string, status: number) {
    super(message);
    this.status = status;
  }
}

/** The file is a trace, but not one this reader knows how to read. */
export class TraceSchemaMismatch extends Error {}

export async function loadTrace(url: string, signal: AbortSignal): Promise<Trace> {
  // The signal reaches the network stack rather than only a flag, because this
  // is an 8-9 MB body: a reader who leaves the arena one second in would
  // otherwise go on downloading a fight for a route that no longer exists, and
  // a second visit inside that window would run a second download beside it.
  const response = await fetch(url, { signal });
  if (response.status === 404) {
    throw new TraceUnavailable(`${url}: ${response.status} ${response.statusText}`, 404);
  }
  // A host that rewrites unknown paths to its own index page reports the same
  // absence as 200 and an HTML body -- `vite preview` does it, and so does most
  // static hosting. Read as a parse failure it would blame a missing file for
  // being corrupt; read here as what it is, it reaches the same explanation a
  // 404 does.
  if (response.ok && (response.headers.get("content-type") ?? "").includes("html")) {
    throw new TraceUnavailable(
      `${url}: the server answered with a page rather than a recording`, response.status,
    );
  }
  if (!response.ok) {
    throw new Error(`${url}: ${response.status} ${response.statusText}`);
  }
  const value = (await response.json()) as Trace;
  // Refused rather than adapted to. A viewer that guesses at an unknown schema
  // is the failure this whole detour exists to avoid: being shown something
  // plausible that is not what the simulation did.
  if (value.schema !== TRACE_SCHEMA) {
    throw new TraceSchemaMismatch(
      `trace schema is ${String(value.schema)}, this page reads ${TRACE_SCHEMA}`,
    );
  }
  if (!Array.isArray(value.frames) || value.frames.length === 0) {
    throw new Error("trace carries no frames");
  }
  return value;
}

export function add(a: V3, b: V3): V3 {
  return [a[0] + b[0], a[1] + b[1], a[2] + b[2]];
}

export function sub(a: V3, b: V3): V3 {
  return [a[0] - b[0], a[1] - b[1], a[2] - b[2]];
}

export function scale(a: V3, k: number): V3 {
  return [a[0] * k, a[1] * k, a[2] * k];
}

export function length(a: V3): number {
  return Math.hypot(a[0], a[1], a[2]);
}

/**
 * The four corners of a shield face, in the order `shield_face` builds them.
 *
 * **The one piece of simulation geometry this page rebuilds**, mirroring
 * `crates/sim/src/combat/geometry.rs:119`. It is not published: the collider is
 * crate-private and the pose row carries the centre, the normal and the extents
 * instead. The mirror is four lines and it is drawn together with its own centre
 * and normal, so a drift shows up on screen as a rectangle that has come off its
 * own marker rather than as a quietly wrong picture.
 */
export function shieldCorners(shield: ShieldFace): readonly [V3, V3, V3, V3] {
  const front = add(shield.centre, scale(shield.normal, shield.thickness / 2 / 65536));
  const left: V3 = [-shield.normal[1], shield.normal[0], 0];
  const side = scale(left, shield.halfWidth / 65536);
  const up: V3 = [0, 0, shield.halfHeight];
  return [
    sub(sub(front, side), up),
    sub(add(front, side), up),
    add(add(front, side), up),
    add(sub(front, side), up),
  ];
}

File: crates/lab/src/trace.rs
//! One fight, written down so somebody can look at it.
//!
//! **This is a diagnostic artifact and it is deliberately not the worker
//! protocol.** v2-17 closed with the mechanical gate failing by roughly a factor
//! of fifty and with three successive diagnoses refuted by later measurement,
//! and the closure's first instruction to a successor is to go and watch a fight
//! before calibrating anything else. Building checkpoint C's ABI to do that
//! would pin a published pose row against a physics nobody has looked at yet.
//! So the trace leaves this crate as JSON, is read by a page that talks to no
//! worker and instantiates no wasm, and can be deleted in the commit that lands
//! the real path.
//!
//! **Every number in the file is a raw fixed-point integer.** `Fx` is 16.16, a
//! JavaScript double holds an `i32` exactly and dividing one by 65536 is exact,
//! so a reader recovers the simulation's number rather than a rounding of it.
//! The other half of the reason is that this crate prints no floating point
//! anywhere, and a viewer is not the place to start.
//!
//! Nothing here computes geometry. The regional capsules come from
//! [`sim::body_region_volumes`], which is the same function the contact phase
//! sweeps, precisely so that the picture cannot disagree with the physics about
//! where an arm is. A viewer that rebuilt a shoulder from an anatomy row would
//! be a second answer to that question, and the one thing this detour exists to
//! avoid is being confidently shown the wrong thing.

use fx::Vec3;
use sim::{
    AnatomyRegion, BodyAnatomySpec, EquipmentGeometry, Faction, Outcome, Scenario, World,
    CONTACT_ENERGY_FLOOR, IMPACT_THRESHOLD,
};
use std::fmt::Write as _;

/// The reader's contract.
///
/// Bumped when a field changes meaning, never when one is added: the page
/// refuses a file whose schema it does not know, which is the difference between
/// "the viewer is out of date" and "the viewer is drawing the wrong thing".
///
/// **Moved from 5 by the forearm collider, and it is a meaning change rather
/// than an addition.** `pose.regions` went from five rows to seven, and until it
/// did, row `i` was named by `regionNames[i]`. It no longer is: the array is the
/// swept-volume list and the name table is anatomy, which agree on their first
/// five entries and nowhere after. A schema-5 file read by this page would draw
/// two arms as one capsule each and would fail
/// `a_live_fight_matches_the_traced_fight` deep inside a pose diff -- which is
/// the confusing half of the failure the guard exists to make plain.
pub const TRACE_SCHEMA: &str = "arpg-fight-trace-6";

/// Raw units in one world unit, carried in the file rather than assumed by the
/// reader. `Fx::ONE.raw()` is not public API to a page.
const ONE_RAW: i32 = 1 << 16;

/// What the header says about the run, gathered where the run was driven.
///
/// A struct rather than seven parameters because six of the seven are already
/// columns of the trial the gate produced, and a positional argument list of
/// `(seed, script, mirrored, timed_out, ...)` is two `bool`s away from being
/// silently transposable.
pub struct TraceRun<'a> {
    pub scenario: &'a Scenario,
    pub seed: u64,
    /// What drove the Fighter, and what drove the Brute.
    ///
    /// **Two fields where schema 2 had one `script`, and the reason is v2-19.**
    /// Until a learned policy existed, every trace put the same script on both
    /// sides and one token said everything; a learned fight is a fight with a
    /// different policy on each side, and a viewer showing "learned" over a run
    /// whose opponent was the windmill rather than the composed script would be
    /// showing something plausible that is not what happened -- which is the one
    /// failure this whole detour exists to avoid.
    pub heroes: &'a str,
    pub monsters: &'a str,
    /// The SHA-256 of the checkpoint that produced the learned side, or `None`.
    ///
    /// The name of the weights and not the weights: a trace is a recording of a
    /// fight, and a reader that wanted to re-run it needs the checkpoint file,
    /// which this identifies and does not contain.
    pub checkpoint: Option<&'a str>,
    pub mirrored: bool,
    pub outcome: Outcome,
    pub timed_out: bool,
    pub ticks: u32,
}

/// Frames accumulated as they happen, serialised on arrival.
///
/// Held as text rather than as a `Vec` of rows because there is no consumer for
/// a row in this process: the whole artifact is written once and read by a
/// browser. A 3600-tick fight is about six megabytes of it, which is a file, not
/// a data structure.
pub struct FightTrace {
    /// Anatomy per unit, in `Scenario::units` order. Resolved once at
    /// construction so a per-tick binary search does not sit inside the loop
    /// that the gate's own measurement shares.
    anatomy: Vec<BodyAnatomySpec>,
    /// `(kind, faction)` per unit, so the slot assumption below is checked
    /// rather than believed.
    identity: Vec<(sim::Body, sim::Faction)>,
    frames: String,
    recorded: u32,
    limit: u32,
}

impl FightTrace {
    /// `limit` bounds the **recording**, not the fight.
    ///
    /// The run always goes to its natural stop, because the trial this recorder
    /// hangs off is the one the gate reports and truncating the world would make
    /// the two disagree. A truncated file says so in its header, so a viewer
    /// showing 600 of 3600 ticks can never be mistaken for a fight that ended.
    pub fn new(scenario: &Scenario, limit: u32) -> FightTrace {
        let table = scenario.combat_specs.as_ref()
            .expect("an articulated scenario carries a combat spec table");
        let anatomy = scenario.units.iter().map(|unit| {
            let row = unit.articulated.expect("an articulated unit carries a spec row");
            table.anatomy(row.anatomy).expect("a validated anatomy reference").clone()
        }).collect();
        let identity = scenario.units.iter().map(|unit| (unit.kind, unit.faction)).collect();
        FightTrace { anatomy, identity, frames: String::new(), recorded: 0, limit }
    }

    /// One frame: every published pose, every resolution row the tick produced,
    /// and the two health fractions the outcome is decided on.
    ///
    /// Called once before the first step and once after every step, so frame `t`
    /// is the world as tick `t` left it and frame 0 is the fixture as it spawned.
    pub fn record(&mut self, world: &World) {
        if self.recorded >= self.limit {
            return;
        }
        if self.recorded > 0 {
            self.frames.push(',');
        }
        self.recorded += 1;

        let out = &mut self.frames;
        // `fmt::Write` on a `String` cannot fail; there is nothing for the
        // `Result` to say. Discarded once per call site rather than unwrapped,
        // because a panic path here would be unreachable code in a diagnostic.
        let _ = write!(out, "{{\"t\":{},\"poses\":[", world.tick());

        for (n, pose) in world.articulated_poses().enumerate() {
            if n > 0 {
                out.push(',');
            }
            // A world slot indexes the unit that spawned into it, which is true
            // because `World::try_new` spawns `scenario.units` in order and this
            // fixture never respawns -- and is an assumption about spawn order
            // rather than a documented contract, so it is checked. A trace that
            // hung the Brute's anatomy off the Fighter would draw two bodies
            // that were both plausible and both wrong.
            let slot = pose.id.index as usize;
            let anatomy = self.anatomy.get(slot)
                .expect("a pose slot indexes the unit that spawned into it");
            if let (Some(&(kind, faction)), Some(view)) =
                (self.identity.get(slot), world.view(pose.id))
            {
                assert!(view.kind == kind && view.faction == faction,
                    "slot {slot} holds a {:?} of {:?}, not the {kind:?} of {faction:?} that \
                     scenario.units[{slot}] describes", view.kind, view.faction);
            }
            // Presence per region, from the one authority on it. A severed arm
            // has no capsule to draw and is not a zero-radius point.
            let present: [bool; AnatomyRegion::COUNT] =
                std::array::from_fn(|part| pose.severed_mask & (1 << part) == 0);
            // The published hands are world space and `body_region_volumes`
            // takes them body-relative -- it adds the origin itself, which is
            // the single conversion the whole contact module is arranged around.
            let hands = [pose.arms[0].hand - pose.body, pose.arms[1].hand - pose.body];
            // The **swept volumes**, all seven of them, through the same
            // constructor `crates/web` publishes with. A jointed arm is two
            // capsules, and `a_live_fight_matches_the_traced_fight` compares
            // this array against the live one field for field -- so a trace that
            // wrote the five anatomy regions here would fail that comparison
            // rather than merely draw a shorter arm. The elbows are `None` on
            // every articulated body, which is why no recorded fixture's rows 5
            // and 6 are anything but absent.
            let elbows = [pose.arms[0].elbow.map(|joint| joint - pose.body),
                          pose.arms[1].elbow.map(|joint| joint - pose.body)];
            let regions = sim::jointed_body_region_volumes(
                pose.body, anatomy, pose.body_yaw, hands, present, elbows,
            );

            let _ = write!(out, "{{\"id\":");
            entity(out, pose.id);
            let _ = write!(out, ",\"body\":");
            vec3(out, pose.body);
            let _ = write!(out, ",\"yaw\":{},\"vel\":", pose.body_yaw.raw());
            vec3(out, pose.body_velocity);

            out.push_str(",\"arms\":[");
            for (limb, arm) in pose.arms.iter().enumerate() {
                if limb > 0 {
                    out.push(',');
                }
                out.push_str("{\"hand\":");
                vec3(out, arm.hand);
                out.push_str(",\"vel\":");
                vec3(out, arm.velocity);
                out.push_str(",\"target\":");
                vec3(out, arm.target_hand);
                let _ = write!(out, ",\"fatigue\":{}}}", arm.fatigue.raw());
            }

            out.push_str("],\"weapons\":[");
            for (limb, weapon) in pose.weapons.iter().enumerate() {
                if limb > 0 {
                    out.push(',');
                }
                match weapon {
                    None => out.push_str("null"),
                    Some(segment) => {
                        out.push_str("{\"hilt\":");
                        vec3(out, segment.hilt);
                        out.push_str(",\"tip\":");
                        vec3(out, segment.tip);
                        let _ = write!(out, ",\"radius\":{}}}", segment.radius.raw());
                    }
                }
            }

            out.push_str("],\"shield\":");
            match pose.shield {
                None => out.push_str("null"),
                Some(shield) => {
                    out.push_str("{\"centre\":");
                    vec3(out, shield.centre);
                    out.push_str(",\"normal\":");
                    vec3(out, shield.normal);
                    let _ = write!(out, ",\"halfWidth\":{},\"halfHeight\":{},\"thickness\":{}}}",
                        shield.half_width.raw(), shield.half_height.raw(), shield.thickness.raw());
                }
            }

            out.push_str(",\"regions\":[");
            for (at, region) in regions.iter().enumerate() {
                if at > 0 {
                    out.push(',');
                }
                out.push_str("{\"lower\":");
                vec3(out, region.lower);
                out.push_str(",\"upper\":");
                vec3(out, region.upper);
                let _ = write!(out, ",\"radius\":{},\"present\":{}}}",
                    region.radius.raw(), region.present);
            }

            out.push_str("],\"integrity\":");
            fractions(out, &pose.integrity_fraction);
            out.push_str(",\"wound\":");
            fractions(out, &pose.wound_fraction);
            let _ = write!(out, ",\"blood\":{},\"shock\":{},\"severed\":{},\"equipmentMask\":{}",
                pose.blood_fraction.raw(), pose.shock.raw(),
                pose.severed_mask, pose.equipment_mask);

            let (intent, target) = match pose.intent {
                sim::Intent::Hold => ("hold", None),
                sim::Intent::Flee => ("flee", None),
                sim::Intent::Attack(id) => ("attack", Some(id)),
            };
            let _ = write!(out, ",\"intent\":\"{intent}\",\"target\":");
            match target {
                None => out.push_str("null"),
                Some(id) => entity(out, id),
            }
            let _ = write!(out, ",\"hints\":[{},{}]}}",
                pose.hints[0] as u8, pose.hints[1] as u8);
        }

        out.push_str("],\"projectiles\":[");
        for (n, projectile) in world.articulated_projectiles().enumerate() {
            if n > 0 {
                out.push(',');
            }
            let _ = write!(out, "{{\"id\":[{},{}],\"owner\":",
                projectile.slot, projectile.generation);
            entity(out, projectile.owner);
            out.push_str(",\"position\":");
            vec3(out, projectile.position);
            out.push_str(",\"velocity\":");
            vec3(out, projectile.velocity);
            let _ = write!(out, ",\"radius\":{},\"remainingRange\":{}}}",
                projectile.radius.raw(), projectile.remaining_range.raw());
        }

        out.push_str("],\"contacts\":[");
        for (n, row) in world.contact_resolutions().iter().enumerate() {
            if n > 0 {
                out.push(',');
            }
            let key = row.fact.key;
            out.push_str("{\"a\":");
            entity(out, key.a);
            let _ = write!(out, ",\"aSlot\":{},\"b\":", key.a_slot);
            entity(out, key.b);
            let _ = write!(out, ",\"bSlot\":{},\"kind\":{},\"region\":{}",
                key.b_slot, key.kind as u8,
                // The **region**, mapped, and not the volume the fact carries.
                // `regionNames` below has five entries and `Trace.region` is
                // documented as an index into it, so a forearm is written as its
                // arm. Writing the raw volume would put a `5` in a column the
                // reader names from a five-element table.
                sim::volume_region(row.fact.volume as usize)
                    .map_or(sim::NO_VOLUME, |part| part as u8));
            out.push_str(",\"point\":");
            vec3(out, row.fact.point);
            out.push_str(",\"normal\":");
            vec3(out, row.fact.normal);
            out.push_str(",\"velocityA\":");
            vec3(out, row.fact.velocity_a);
            out.push_str(",\"velocityB\":");
            vec3(out, row.fact.velocity_b);
            out.push_str(",\"impulseA\":");
            vec3(out, row.impulse.on_a);
            out.push_str(",\"impulseB\":");
            vec3(out, row.impulse.on_b);
            // **The ledger is named `group` because it is one.**
            // `resolve_group_into` builds a single `EnergyLedger` per time-of-
            // impact group and copies it into every row of that group
            // (`resolution.rs:267` and `:282`), so five facts in one group carry
            // five copies of one number -- and that number is
            // `closure_energy` over every collider in the group, bodies' own
            // translational energy included. It is not what any one contact
            // brought and it is not what the floor is charged against. A reader
            // that plotted it against `CONTACT_ENERGY_FLOOR` would be comparing
            // a group's whole kinetic energy to a per-fact deduction, and on
            // this corpus that reads 578 contacts over the floor where the
            // honest count is 9.
            //
            // The per-fact quantity is `share`, the row's slice of
            // `dissipated`, and it is recoverable exactly:
            // `cut + thrust + pressure == share` for every weapon-body row. For
            // the other two kinds no channel exists and no floor is charged.
            //
            // **That invariant is why `crush_raw` is added in rather than
            // dropped, and this line published `pressure_raw` alone until it was
            // caught.** `channels()` splits a share four ways --
            // `(cut, thrust, crush, pressure)` -- while this row, like the
            // combat-event ABI beside it, carries three channel words. The
            // comment here described a three-way `channels()` that had already
            // become four, so the missing term read as an absent one. Publishing
            // only `pressure_raw` is invisible while `crush_raw` is zero, which
            // is every default-law fixture, and wrong the moment one is not: on
            // seed 3 of the shipped duel, frame 460's first contact splits 194
            // into crush 36 and pressure 157, and a reader checking the
            // invariant against the recorded columns finds 158. `crates/web/
            // src/lib.rs`'s `COMBAT_EVENT_PRESSURE` word already summed the two
            // for exactly this reason; the two sides of the trace contract now
            // agree, which is what `a_live_fight_matches_the_traced_fight`
            // compares field for field.
            let _ = write!(out,
                ",\"toi\":{},\"group\":{},\"alpha\":{},\
                 \"groupBefore\":{},\"groupAfter\":{},\"groupDissipated\":{},\
                 \"cut\":{},\"thrust\":{},\"pressure\":{},\"deflected\":{},\"severed\":{}}}",
                row.fact.toi.get().raw(), row.group_ordinal, row.group_alpha_raw,
                row.energy.before_raw, row.energy.after_raw, row.energy.dissipated_raw,
                row.cut_raw, row.thrust_raw, row.crush_raw + row.pressure_raw,
                row.deflected_raw, row.severed);
        }

        let _ = write!(out, "],\"health\":[{},{}]}}",
            world.health_fraction(Faction::Heroes).raw(),
            world.health_fraction(Faction::Monsters).raw());
    }

    /// The complete file. Consumes the recorder: there is exactly one artifact.
    pub fn finish(self, run: &TraceRun) -> String {
        let scenario = run.scenario;
        let arena = scenario.arena();
        let mut out = String::with_capacity(self.frames.len() + 4096);

        let _ = write!(out, "{{\"schema\":\"{TRACE_SCHEMA}\",\"one\":{ONE_RAW}");
        let _ = write!(out, ",\"scenario\":\"{}\",\"mirrored\":{}", scenario.name, run.mirrored);
        // **What was run, not whether it was the pin.** This field used to be
        // `null` for every mirrored run, on the argument that the only question a
        // reader asks of it is "am I looking at the pinned fixture" -- true while
        // the only two scenarios a trace could record were the fixture and its
        // reflection. A described duel killed that premise: a mirrored
        // `configured-duel-v1` is one of unboundedly many scenarios, its
        // fingerprint is the only thing that says *which*, and the file is the
        // artifact that outlives the terminal the number was printed in. So the
        // scenario's own number goes in, always, and a reader deciding whether
        // it holds the pin compares it against the pin -- which is the same test
        // it was already making, one comparison later. `mirrored` is a separate
        // field and still says a reflection happened.
        let _ = write!(out, ",\"fingerprint\":\"{:#018x}\"", scenario.fingerprint());
        let _ = write!(out, ",\"seed\":{},\"heroes\":\"{}\",\"monsters\":\"{}\"",
            run.seed, run.heroes, run.monsters);
        match run.checkpoint {
            None => out.push_str(",\"checkpoint\":null"),
            Some(digest) => {
                let _ = write!(out, ",\"checkpoint\":\"{digest}\"");
            }
        }
        let _ = write!(out, ",\"outcome\":\"{:?}\",\"timedOut\":{},\"ticks\":{}",
            run.outcome, run.timed_out, run.ticks);
        let _ = write!(out, ",\"maxTicks\":{},\"arena\":[{},{}]",
            scenario.max_ticks, arena.x.raw(), arena.y.raw());
        let _ = write!(out, ",\"frameCount\":{},\"truncated\":{}",
            self.recorded, self.recorded < run.ticks + 1);

        // The two lines the picture is judged against, read out of the code that
        // owns them. `IMPACT_THRESHOLD` is the legacy "is this a swing at all"
        // bar and is not what the articulated path tests -- it is carried
        // because it is the only calibrated speed in the repository, and the
        // finding that closed v2-17 is that a weapon-body contact arrives about
        // thirty-five times under it.
        let _ = write!(out, ",\"impactThreshold\":{},\"contactEnergyFloor\":{}",
            IMPACT_THRESHOLD.raw(), CONTACT_ENERGY_FLOOR);

        // Names for every code the frames carry as a number, so the reader owns
        // no copy of an enum that `sim` can renumber underneath it.
        // Five names for a `regions` array that is seven long, and the
        // mismatch is the point rather than a bug: this table names **anatomy**
        // -- it is what the contact column and the severed readout index -- while
        // `pose.regions` is the swept-volume list the solver works in. The first
        // five agree by construction, which is what `forearm_volume` appending
        // rather than interleaving buys.
        out.push_str(",\"regionNames\":[\"head\",\"torso\",\"leftArm\",\"rightArm\",\"legs\"]");
        out.push_str(",\"hintNames\":[\"idle\",\"chasing\",\"braced\",\"contact\",\"recoiling\",\"severed\"]");
        out.push_str(",\"contactKinds\":[\"weaponWeapon\",\"weaponShield\",\"weaponBody\",\"projectileBody\"]");
        let _ = write!(out, ",\"bodySlot\":{},\"noRegion\":{}", sim::BODY_SLOT, sim::NO_VOLUME);

        out.push_str(",\"bodies\":[");
        for (index, unit) in scenario.units.iter().enumerate() {
            if index > 0 {
                out.push(',');
            }
            let anatomy = self.anatomy.get(index).expect("one anatomy per unit");
            let _ = write!(out, "{{\"index\":{index},\"kind\":\"{:?}\",\"faction\":\"{:?}\"",
                unit.kind, unit.faction);
            let _ = write!(out,
                ",\"anatomy\":{{\"standingHeight\":{},\"shoulderHeight\":{},\
                 \"shoulderHalfWidth\":{},\"armLength\":{},\"handRadius\":{}}}",
                anatomy.standing_height.raw(), anatomy.shoulder_height.raw(),
                anatomy.shoulder_half_width.raw(), anatomy.arm_length.raw(),
                anatomy.hand_radius.raw());

            out.push_str(",\"carried\":[");
            let table = scenario.combat_specs.as_ref().expect("a validated spec table");
            let row = unit.articulated.expect("a validated articulated unit");
            for (slot, id) in row.equipment.iter().enumerate() {
                if slot > 0 {
                    out.push(',');
                }
                match id.and_then(|id| table.equipment(id)) {
                    None => out.push_str("null"),
                    Some(item) => {
                        let _ = write!(out, "{{\"action\":\"{:?}\",\"binding\":\"{:?}\",\
                            \"mass\":{},\"balance\":{}",
                            item.action, item.binding, item.mass.raw(), item.balance.raw());
                        match item.geometry {
                            EquipmentGeometry::Segment { length, radius } => {
                                let _ = write!(out, ",\"geometry\":\"segment\",\
                                    \"length\":{},\"radius\":{}}}", length.raw(), radius.raw());
                            }
                            EquipmentGeometry::Shield { half_width, half_height, thickness } => {
                                let _ = write!(out, ",\"geometry\":\"shield\",\"halfWidth\":{},\
                                    \"halfHeight\":{},\"thickness\":{}}}",
                                    half_width.raw(), half_height.raw(), thickness.raw());
                            }
                        }
                    }
                }
            }
            out.push_str("]}");
        }

        out.push_str("],\"frames\":[");
        out.push_str(&self.frames);
        out.push_str("]}\n");
        out
    }
}

fn vec3(out: &mut String, v: Vec3) {
    let _ = write!(out, "[{},{},{}]", v.x.raw(), v.y.raw(), v.z.raw());
}

fn entity(out: &mut String, id: sim::EntityId) {
    let _ = write!(out, "[{},{}]", id.index, id.generation);
}

fn fractions(out: &mut String, values: &[fx::Fx]) {
    out.push('[');
    for (at, value) in values.iter().enumerate() {
        if at > 0 {
            out.push(',');
        }
        let _ = write!(out, "{}", value.raw());
    }
    out.push(']');
}
